### PR TITLE
Add PT100 on MCU's with 3.3v excitation temptable

### PR DESCRIPTION
**Pull request by petrzmax**
### Description

I was trying to set up PT100 sensor which e3d amp board on my mks robin nano, but I had some problems. Being 100% sure that wiring is ok I was getting about 180 deg at room temperature while on the signal pin a had 1.2V. I was trying different pins and even removed the pull up resistor but still getting same results. I was searching a bit and i have found this comment: https://github.com/MarlinFirmware/Marlin/issues/4987#issuecomment-526885579

@tphan26363 comment sounded logical, so I have recalculated existing temptable. And it worked like a charm! Now I'm getting right temperature.

tl;dr It's probably related to different pullup voltage (3.3 / 5v) so You may want to use different name etc. in the commented sections.

### Benefits
Easily use PT100 with amp board on MCU's with 3.3v excitation (STM32 \ LPC176x....)

### Related Issues

https://github.com/MarlinFirmware/Marlin/issues/4987

### CODE REVIEW

**Comment by InsanityAutomation** [comment]
While in essence I agree, I would consider the naming a bit. The root issue is the excitation, 3.3v or 5v. I would label it as such as some boards may be 5v with a 32 bit mcu.

**Comment by petrzmax** [comment]
@InsanityAutomation You are right my friend. I'm a bit lame in giving right names. Do You have some 
 idea how I should name it?

(A bit new to open source projects question) should I force push ammended commit with fixed names and message or just make another one to my branch?

**Comment by InsanityAutomation** [comment]
just another push to youre branch is fine, doesnt need to be a force push. I would just state excitation voltage on each one

```
*    20 : Pt100 with circuit in the Ultimainboard V2.x with 5v excitation (AVR)
 *    21 : Pt100 with circuit in the Ultimainboard V2.x with 3.3v excitation (STM32 \ LPC176x....)
```

As the rest of the circuit with the amp board is still the same. Add in any processors you know for sure are 3.3v.

**Comment by petrzmax** [comment]
@InsanityAutomation Thank You my friend. Pushed it a moment ago ;)

**Comment by InsanityAutomation** [comment]
Sounds good, just the termistor names file needs a corresponding edit :) After that up to Thinky!

**Comment by petrzmax** [comment]
These should be short I think, so I have added just voltage there ;) I hope it's all good to go now :D

**Comment by thinkyhead** [comment]
@InsanityAutomation — Do you anticipate we will want to do this for all the tables? And if so, can we just automate this by adding a factor that gets applied to all tables based on the board's logic level?

**Comment by InsanityAutomation** [comment]
@thinkyhead no, because most have the thermistor setup through a level adjusted with the sense resistor. This is different because it doesnt go through the normal circuit prepared for it. It goes to a raw analog input.

**Comment by thinkyhead** [comment]
And yet the conversion of table 20 using the formula `N * 5 / 3.3` lines up exactly with the values in table 21. Seems like we could use this for some cases instead of defining separate tables.

**Comment by thinkyhead** [comment]
Just for kicks, I tweaked the thermistor table to expose the relationship between it and table 20, and introducing a method to allow for synthesizing other tables in the future.

**Comment by InsanityAutomation** [comment]
Perhaps an extension to type 1000 p parameter to apply a multiplier would be a nice wishlist item.

**Comment by Bob-the-Kuhn** [comment]
I'm confused about what problem you are trying to solve.

In my experience the thermistor table does NOT change if the Vref for the ADC is the same as the excitation to the PT100 circuit.  It's the RATIO of Vin to Vmax that is important.  That ratio is only dependent on the resistance values of the sensor and the pullup resistor. 

NOTE - the PT100 amplifier is more complicated but it is still essentially a pullup resistor tied to the sensor.  The amplifier's output (at any given temperature) changes directly proportional to the change in the power supplied to the amplifier. The RATIO of Vin to Vmax remains the same.

For example, a PT100 with a 4.7K pullup will use table 147 if it's on an AVR board or a 32 bit board.  A PT100 amplifier system will use table 20 if it's on an AVR board or a 32 bit board.  

The same is true if using a PT100 amplifier. On an AVR system the amplifier needs to be given a 5V supply.  On a 32 bit board it needs to be given a 3.3V supply.   In **no case** should the PT100 amplifier be powered from 5V on a 32 bit system (most analog inputs are not 5V tolerant).  In **no case** should a pullup resistor be used with a PT100 amplifier.   I've had one PT100 amplifier that worked only on a 5V (AVR) system and another that worked on either 5V or 3.3V systems.

In my experience a PT100 system being off by 5-10C at room temperature is usually caused by the ground reference for the PT100 being at a higher voltage (as in 10-20mV) due to current flowing in the thin ground traces on the controller board.  It is good practice to run a heavier wire between the analog input port's ground pin and the main ground pin on the controller's power connector.  Some improvement is also seen if an additional ground wire can be attached to a ground near the CPU.

If you absolutely must attach the PT100 to an analog input with a pullup resistor then my recommendation is to ditch the PT100 amplifier and wire the PT100 sensor directly to the port.  

A custom thermistor table is needed if a PT100 amplifier is connected to a port with a pullup resistor.  There's enough variation from amplifier to amplifier that a table customized to that amplifier/controller is a good idea.  I have an excel spreadsheet that will generate the table given a couple of voltage measurements.  

---

I suspect you have a wiring error or a bad PT100 amplifier or are connected to a port with a pullup resistor.

Please provide a drawing of how your system is setup.   It must include details on what each pin on the PT100 amplifier's output connector is tied to.

**Comment by thinkyhead** [comment]
Since the table 21 values are exactly (5.0 / 3.3) times larger than the table 20 values it implies that the signal coming from the device is 5V but the ADC input is expecting current to be relative to 3.3V. If that is the case, I'm not sure how long a 3.3V input will enjoy that.

**Comment by petrzmax** [comment]
@Bob-the-Kuhn I'm working on mks robin nano at the moment so I'm basing on its schematic etc. Here are some informations about this board so we can figure this out:
Endstop has 3 pins: signal, ground and 5V, so all this events should be prepared for 5V. 
btw, I have removed R56 resistor.
I'm not sure how it's handled by mcu, but a lot of additional "things" are working on 5V. Also I think that ADC converter working with 3.3v logic should give me max reading at my connected 5V logic amp but it's not doing this. So If this is normal behaviour for this board I'm not worried about breaking something.

Here are schematics:






As You can see TH1 is connected similiar to X endstop. The difference is that there is HGND and DGND used.  
HGND = power ground (High Voltage)
DGNF = digital ground or down (I'm not sure)

Here are all schamatics if You want to check this out.
[MKS Robin Nano V1.1_001 SCH.pdf](https://github.com/MarlinFirmware/Marlin/files/4139604/MKS.Robin.Nano.V1.1_001.SCH.pdf)

**Comment by tphan26363** [comment]
Just let you know that if you look at the data sheet of some of the chip,  The ADC input can withstand +4 above VDD.  For example: 
https://www.st.com/resource/en/datasheet/stm32f405rg.pdf

**Comment by Bob-the-Kuhn** [comment]
The max voltage for an analog input is 4.0V.  Above that a parasitic diode to 3.3V gets turned on.  It takes some looking through the data sheet but here are the relavent passages:
- Table 11:    `VIN    5 V tolerant except when in analog mode ...  Input voltage on any other pin     4.0 MAX`
- Table 7 note 4:`    5 V tolerant except when in analog mode`

It's safest to power the PT100 amplifier from 3.3V.  That way you're guaranteed to not cause any problems.

On the other hand the PT100 amplifier has a VERY weak output and you'd need to be over 800C to get to 4V (or have the PT100 amplifier's input open) so I doubt that there's been any damage done.

**Comment by InsanityAutomation** [comment]
This would go back to a thought we discussed in passing a year or so ago of putting a max temp sanity check in each thermistor table. Could protect from damage in this situation.

Just in case anyone planned to fdm mig wire or something :)

**Comment by petrzmax** [comment]
@Bob-the-Kuhn Thank You, now everything is clear :)
Have a good day.

**Comment by goldjunge91** [comment]
Hello, sorry for my bad english but in my case the pt100 with e3d amp board don't work on my skr1.3 i connect it like this picture and make in my config 21 read 1 degree only  and when i plug this with same settings on 3.3v i have the same issues <a href="https://www.xup.in/dl,26439703/73393835-a82d4980-42dc-11ea-8b0d-df971f537b12.jpg/" target="_blank"></a>

**Comment by goldjunge91** [comment]
for me on my skr 1.3 with marlin 2.5.3 work this <a href="https://www.xup.in/dl,11121854/73393833-a82d4980-42dc-11ea-8cb4-317c1b810491.jpg/" target="_blank"></a>
and change in pins_BTT_SKR_common.h the TEMP_0_PIN to #define TEMP_0_PIN P0_02_A7 for the TX0 pin and for the rx0 pin P0_03_A6

i add this in my config maybe you can make something like this for other user 
#if TEMP_SENSOR_0 == 20
  #define TEMP_0_PIN  P0_03_A6
#endif

https://github.com/bigtreetech/BIGTREETECH-SKR-V1.3/issues/212
@bdelia thanks a lot for that you find the pin numbers
